File: service/src/http/response.ts
import express from 'express';

/** Shape of a response object to send to the client */
export interface Response<T> {
  /** The actual response payload */
  payload: T;
  /** The status code */
  status?: number;
  /** The content type */
  contentType?: string;
  /** The period of time to cache for */
  cachePeriod?: number;
  /** The etag to use */
  etag?: string;
}

/**
 * Send a response to the client
 * @param res The express Response object
 * @param response The response to send to the client
 */
export function respond<T>(res: express.Response, response: Response<T>): void {
  if (response.etag !== undefined) {
    if (res.req?.header('if-none-match') === response.etag) {
      res.status(304);
      res.end();
      return;
    } else {
      res.setHeader('etag', response.etag);
    }
  }

  if (response.status !== undefined) {
    res.status(response.status);
  }
  if (response.contentType !== undefined) {
    res.contentType(response.contentType);
  }
  if (response.cachePeriod !== undefined) {
    res.setHeader('cache-control', `public,max-age=${response.cachePeriod}`);
  }
  res.json(response.payload);
}

File: service/src/http/response.rs
use actix_http::{
    http::header::CacheControl,
    http::{
        header::{CacheDirective, ContentType, ETag, EntityTag},
        StatusCode,
    },
    Error, Response as HttpResponse,
};
use actix_web::Responder;
use futures::future::{ok, Ready};
use serde::Serialize;

/// Friendly representation of the HTTP Response to send back.
pub struct Response<T>
where
    T: Serialize,
{
    /// The HTTP Status Code
    pub status: StatusCode,
    /// The caching controls
    pub cache_control: Vec<CacheDirective>,
    /// The content type
    pub content_type: ContentType,
    /// The etag to return, if applicable
    pub etag: Option<EntityTag>,
    /// The actual response body
    pub body: Option<T>,
}

impl<T> Default for Response<T>
where
    T: Serialize,
{
    fn default() -> Self {
        Self {
            status: StatusCode::OK,
            cache_control: vec![CacheDirective::NoCache],
            content_type: ContentType::json(),
            etag: None,
            body: None,
        }
    }
}

impl<T> Responder for Response<T>
where
    T: Serialize,
{
    type Error = Error;
    type Future = Ready<Result<HttpResponse, Error>>;

    fn respond_to(self, _req: &actix_web::HttpRequest) -> Self::Future {
        let mut response = HttpResponse::build(self.status);

        response.set(CacheControl(self.cache_control));
        response.set(self.content_type);

        if let Some(etag) = self.etag {
            response.set(ETag(etag));
        }

        let built = response.json(self.body);

        ok(built)
    }
}
